Add tuple and list membership modules

ax0 opened the pull request:
This PR adds tuple and list membership modules to the `gpcircuits` package. This includes additions to the proto POD GPC and modifications to the tests. All tests pass.

Some relevant notes for context:

* Tuples are represented as Poseidon hashes of arrays of entry value hashes of some fixed length (assumed to be between 2 and 4), and the module outputs multiple tuples as dictated by one of its parameters. Value hashes are chosen by index, and tuples may be fed into other tuples by appropriate indexing. An index of -1 implies a reference to 0, the value of choice for padding. One tuple module is instantiated in the proto POD GPC.

* List membership is checked in the usual linear way. Lists are assumed to be padded with the first element of the list. One list membership module is instantiated in the proto POD GPC and the value whose membership is to be checked is referenced by index, where an index may refer to either an entry value hash or a computed tuple hash. Again, an index equal to -1 implies a reference to 0, though this device isn't strictly necessary in this case.

* Some helper functions were added. A variant of `InputSelector` is included to allow for -1 index trick mentioned above while keeping an out-of-bounds check in place otherwise.

* Some changes were made to the circuit family parameters.

Marking this as a draft for now as the tests need to be fleshed out a bit to cover some edge cases (e.g. the case of an empty membership list), and I'd like to give the code (and comments) a once-over. The design is unlikely to change much, so feel free to peruse.

artwyman commented on packages/lib/gpcircuits/src/proto-pod-gpc.ts:
```diff
@@ -143,6 +163,21 @@ export type ProPODGPCCircuitDesc = CircuitDesc & {
    */
   maxEntries: number;
 
+  /**
```
nit: Why insert these between maxEntries and merkleMaxDepth rather than at the end?
This is somewhat arbitrary, but I prefer to keep these grouped conceptually, and the merkle depth applies to the entries, so I'd rather keep it directly after maxEntries (here and elsewhere like in the circuit names).
We might actually need to rename merkeMaxDepth to podMerkleMaxDepth if/when we have a group membership check with its own merkle proofs.

artwyman commented on packages/lib/gpcircuits/circuits/list-membership.circom:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    if (MAX_LIST_ENTRIES == 0)
+	isMember <== 1;
```
I'd expect 0 here.  No value is a member of an empty list.

artwyman commented on packages/lib/gpcircuits/circuits/list-membership.circom:
```diff
@@ -0,0 +1,36 @@
+    MAX_LIST_ENTRIES
+) {
+    // Value hash to be checked.
+    signal input valueHash; 
```
The name is accurate for our intended use but at this level is there any requirement for this to be a hash?  It's just a value to compare to the list's values.

artwyman commented on packages/lib/gpcircuits/circuits/tuple.circom:
```diff
@@ -0,0 +1,64 @@
+    MAX_TUPLES
+) {
+    // Value hashes as input signals
+    signal input valueHash[MAX_VALUES];
```
Same as in the list module, do these specifically need to be hashes, or are they just inputs where anything can work?

artwyman commented on packages/lib/gpcircuits/circuits/tuple.circom:
```diff
@@ -0,0 +1,64 @@
+    // Maximum number of value hashes to choose from
+    MAX_VALUES,
+    // Maxium number of tuples to generate
+    MAX_TUPLES
```
This is interesting.  What's the value of generating multiple tuples from a single module vs. having multiple tuple modules?

artwyman replied:
Or maybe a MultiTupleModule template to deal with the routing of inputs to outputs between several TupleModule templates in a row.  I think some sort of hierarchical organization would be clearer than the multi-dimensional arrays here.

artwyman commented on packages/lib/gpcircuits/circuits/tuple.circom:
```diff
@@ -0,0 +1,64 @@
+    // Array indicating which values form which tuples by index,
+    // where an index `i` less than MAX_VALUES refers to
+    // `valueHash[i]` and one less than `MAX_VALUES + MAX_TUPLES`
+    // refers to `valueHash[i - MAX_VALUES]`.
```
I don't follow what this is describing.

artwyman commented on packages/lib/gpcircuits/circuits/tuple.circom:
```diff
@@ -0,0 +1,64 @@
+    // Note that tupleIndices[i][j] = -1 refers to 0 by convention.
+
+    // Input selector component.
+    component inputSelector[MAX_TUPLES][TUPLE_ARITY];
```
Is this pre-declaration required, or can we get away with anonymous templates?  That's relatively new in circom, but it simplified a lot of the InputSelectors in the proto-pod-gpc circuit.

artwyman commented on packages/lib/gpcircuits/circuits/tuple.circom:
```diff
@@ -0,0 +1,64 @@
+    signal output tupleHashes[MAX_TUPLES];
+
+    for(var i = 0; i < MAX_TUPLES; i++) {
+	for(var k = 0; k < TUPLE_ARITY; k++) {
```
i, j, k being out of order here is confusing.  Is there a reason the name order is important?
My alternative suggestion would be to give these descriptive names like `tupleIndex` instead.

artwyman commented on packages/lib/gpcircuits/circuits/proto-pod-gpc.circom:
```diff
@@ -28,6 +30,15 @@ template ProtoPODGPC (
+    MAX_TUPLES,
+
+    // Indicates the arity of the tuples.
+    ARITY,
```
TUPLE_ARITY?
Actually I'd suggest maybe using TUPLE_SIZE.  As a non-mathematician remind myself what "arity" meant.  It's correct, just not a commonly-used term in software, in my experience.

artwyman commented on packages/lib/gpcircuits/circuits/proto-pod-gpc.circom:
```diff
@@ -140,13 +151,44 @@ template ProtoPODGPC (
+     * 1 TupleModule with its inputs & outputs.
+     */
+
+    signal input tupleIndices[MAX_TUPLES][ARITY];
```
This needs a comment here describing how to use it.

artwyman commented on packages/lib/gpcircuits/circuits/proto-pod-gpc.circom:
```diff
@@ -140,13 +151,44 @@ template ProtoPODGPC (
+    signal tupleHashes[MAX_TUPLES] <== TupleModule(ARITY, MAX_ENTRIES, MAX_TUPLES)(entryValueHashes, tupleIndices);
+
+    /*
+     * 1 ListMembershipModule with its inputs & outputs.
```
It would be good for this to be scalable to multiple modules as well.

artwyman commented on packages/lib/gpcircuits/circuits/proto-pod-gpc.circom:
```diff
@@ -140,13 +151,44 @@ template ProtoPODGPC (
+     * 1 ListMembershipModule with its inputs & outputs.
+     */
+
+    // Index of entry value (if less than `MAX_ENTRIES`) or tuple hash that ought to be a member.
```
This scheme needs more description here.
Alternately if the use of this index beyond entries becomes a reusable thing, maybe there should be a block comment somewhere to explain how we lay out the list of possible inputs to further modules, including the special role of -1, the appending of entries, then tuples, then more stuff later.

artwyman commented on packages/lib/gpcircuits/circuits/proto-pod-gpc.circom:
```diff
@@ -140,13 +151,44 @@ template ProtoPODGPC (
+
+    var numEntriesAndTuples = MAX_ENTRIES + MAX_TUPLES;
+
+    if(MAX_LIST_ENTRIES > 0) {
```
So the intent here is to use 0 to represent the absence of this module?
Makes sense, though if/when we scale to multiple lists that would be covered by 0 MAX_LISTS, at which point MAX_LIST_ENTRIES can be constrained to be >0.

artwyman commented on packages/lib/gpcircuits/circuits/proto-pod-gpc.circom:
```diff
@@ -140,13 +151,44 @@ template ProtoPODGPC (
+	signal isMember <== ListMembershipModule(MAX_LIST_ENTRIES)(
+	    memberHash, membershipList);
+
+	isMember === 1;
```
I like that this leaves room to be fed into more complex predicates later.

artwyman commented on packages/lib/gpcircuits/src/list-membership.ts:
```diff
@@ -0,0 +1,131 @@
+
+export type ListMembershipModuleOutputNamesType = ["isMember"];
+
+export class ListMembership {
```
Not important, but note that the pattern of putting helpers inside a class isn't required.  The helper here could probably be a free function.  I used the pattern for ProtoPODGPC because there were a lot of helpers grouped together, and I anticipated that there would someday be multiple families with similar helpers.  I'm pretty new to TypeScript so I don't know how standard that pattern might be.

artwyman commented on packages/lib/gpcircuits/src/list-membership.ts:
```diff
@@ -0,0 +1,131 @@
+export type ListMembershipModuleOutputNamesType = ["isMember"];
+
+export class ListMembership {
+  // TODO: Rework this to allow for other modules that might make use of
```
As we discussed, I think separating out the helpers for tuple generation (one at a time) would make this clearer and more reusable.  I'd imagine maybe 3 separate helpers:
- Something to build a single tuple 
- Something to handle chaining together tuples when necessary (this is probably the most complicated/fiddly part to express in config).
- Something to take a bunch of these hashed values and format them for the ListMemebership module.  Mostly dealing with the padding.

artwyman commented on packages/lib/gpcircuits/src/list-membership.ts:
```diff
@@ -0,0 +1,131 @@
+				paramMaxTuples: number
+			       ): [number, bigint[][], bigint[]] {
+    // Check input
+    if (listIndices.length != membershipLists.length)
```
nit: always put braces around code blocks.  It avoids mistakes when lines are added, and I also find it more readable.

artwyman commented on packages/lib/gpcircuits/src/list-membership.ts:
```diff
@@ -0,0 +1,131 @@
+    listIndices.forEach(i => {if(i >= paramMaxValues) throw new Error("The indices cannot exceed MAX_VALUES.");});
+
+    // Helpers
+    const pad = (arr: any, len: number, padding: any) => arr.concat(Array(len - arr.length).fill(padding));
```
Do these need to be `any`?  I think our lint rules will complain about this.  I certainly prefer to avoid uses of `any`, which turn off most type checking.

artwyman commented on packages/lib/gpcircuits/src/list-membership.ts:
```diff
@@ -0,0 +1,131 @@
+				paramTupleArity: number,
+				paramMaxListEntries: number,
+				paramMaxTuples: number
+			       ): [number, bigint[][], bigint[]] {
```
clarity nit: Giving multiple return values names (by putting them in an object) can help make them self-documenting.

artwyman commented on packages/lib/gpcircuits/src/proto-pod-gpc.ts:
```diff
@@ -346,8 +405,8 @@ export class ProtoPODGPC {
-    ProtoPODGPC.curcuitDescForParams(1, 1, 5, 9556),
-    ProtoPODGPC.curcuitDescForParams(1, 5, 8, 19223),
-    ProtoPODGPC.curcuitDescForParams(3, 10, 8, 45276)
+    ProtoPODGPC.curcuitDescForParams(1, 1, 10, 2, 2, 5, 9556),
```
I'd suggest making sure there's more than one value for each param included in the test cases.  So, a larger list and tuple.  The last entry is meant to be larger than the test cases to make sure all the padding works.

artwyman replied:
Also note the costs here will need to be updated when these circuits are updated.  I copied the `cost` entry from the number of non-linear constraints in the output when building from the command-line.  I foresee that being annoying to do repeatedly, so don't bother until it's time to merge.  Another good argument for the single source of truth you suggested.

artwyman commented on packages/lib/gpcircuits/src/tuple.ts:
```diff
@@ -0,0 +1,81 @@
+   * @throws TypeError if `tupleIndices.length != paramMaxTuples`.
+   * @throws TypeError if `tupleIndices[i].length != paramTupleArity` for any i.
+   */
+  public static tupleHashes(
```
I suggest factoring this out similarly to what I described for the list helper.  Separate the code for building one tuple from the code to link them together.

artwyman commented on packages/lib/gpcircuits/src/tuple.ts:
```diff
@@ -0,0 +1,81 @@
+   *  refers to a POD object value with that index and `i >= paramMaxValues`
+   *  refers to the `i - paramMaxValues`th tuple.
+   * @returns `bigint[]` of tuple hashes.
+   * @throws TypeError if the given tuple arity does not lie within [2,4].
```
I don't think it's necessary to call out all the individual @throws scenarios here.  I don't mind, but it might be hard to keep up-to-date when code changes.  I think it's sufficient to say `@throws TypeError if inputs are out of range`.

Actually shouldn't that be a RangeError?

artwyman commented on packages/lib/gpcircuits/test/list-membership.spec.ts:
```diff
@@ -0,0 +1,63 @@
+  it("should successfully verify list non-membership", async () => {
+    await circuit.expectPass(sampleInput2, sampleOutput2);
+  });
+});
```
Should exercise more sizes, but otherwise this module is pretty simple so doesn't need much more.

artwyman commented on packages/lib/gpcircuits/test/proto-pod-gpc.spec.ts:
```diff
@@ -393,12 +420,32 @@ function makeTestSignals(
+
+  // ?
+  const indexListPairs = [
+    [1, [85n, sigEntryValue[1]].map(x => { return { type: "cryptographic", value: x }; })],
```
This use of .map() for 2-3 constant entries seems questionable to me.  Harder to read than the literals, and not much shorter.

artwyman reviewed (commented):
A few more comments primarily on style/structure/clarity.  I didn't go in depth on correctness of the TypeScript helpers since I know have changes planned.

Note that I'm a stickler for consistency and clarity, so I know my comments can seem nitpicky.  Don't take it personally.  I like where this is going after our conversation, though.  Looking forward for the next round.

artwyman reviewed (changes requested):
Should've marked Request Changes on the last one, to make it clear where the next step lies.

ax0 commented:
Alright, so I think this is ready for review. Some notes:
- Tuples are padded with the first tuple index.
- A disabled list membership check corresponds to `memberIndex = -1` and a list of zeroes. While the witness tests with `MAX_LISTS = 0` pass, the test artifact checks do not, so I have avoided templates with zero array lengths.
- Additional circuit parameters have been added more or less arbitrarily. I am open to suggestions for a more appropriate selection.
- Some of the TS code would belong in the GPC package once #1662 is merged, specifically `processLists` in `src/list-membership.ts`, which would probably end up being `compileProofListMembership` in `gpcCompile.ts`.

artwyman commented:
> A disabled list membership check corresponds to memberIndex = -1 and a list of zeroes. While the witness tests with MAX_LISTS = 0 pass, the test artifact checks do not, so I have avoided templates with zero array lengths.

This is interesting.  Probably something we should fix, though it can come later.  Eventually when we have many optional modules it could become a problem if it's not possible to leave them out.  Do you know that the issue is with 0-length inputs, as opposed to something in the circuit code?  I could imagine some conditionals in the variable declarations which might fix the minimum input size at 1.  Or does circom allow including signal declarations inside of an if statement so some inputs don't exist with some parameters?  Something to experiment with.

artwyman commented on packages/lib/gpcircuits/src/proto-pod-gpc.ts:
```diff
@@ -152,6 +172,61 @@ export type ProtoPODGPCCircuitParams = {
+};
+
+/**
+ * Size parameters for GPC data before processing for ProtoPODGPC circuit.
```
Also I don't see this type referenced anywhere, so maybe you already simplified it out?

ax0 replied:
Deferred to the next PR.

artwyman commented on packages/lib/gpcircuits/src/tuple.ts:
```diff
@@ -0,0 +1,171 @@
+  }
+  if (indices.some((i) => i >= firstTupleIndex)) {
+    throw new RangeError(
+      "The first tuple index must be strictly greater than the given indices."
```
We should think at some point about how to handle tuples manually formed by users in the config.  That'll be relevant when there are more places for a tuple to go than just into lists.  I think it's fine not to worry about it yet, though.

artwyman commented on packages/lib/gpcircuits/src/util.ts:
```diff
@@ -146,3 +159,30 @@ export function array2Bits(boolArray: bigint[]): bigint {
 }
+
+/**
+ * Zips up a list of lists, i.e. forms pairs, triples, ... from
```
arrays not lists, if we're talking about TypeScript types.

artwyman commented on packages/lib/gpcircuits/test/proto-pod-gpc.spec.ts:
```diff
@@ -117,6 +134,20 @@ const sampleInput: ProtoPODGPCInputs = {
   /*PUB*/ entryIsValueEnabled: 59n,
   /*PUB*/ entryIsValueHashRevealed: 21n,
+
+  // Entry constraint modules.
```
Why did this move?

artwyman replied:
Still wondering why Entry constraint module inputs got moved to be in the middle of Entry module inputs.  These are intended to be grouped by module, not by name, as per the comments.

ax0 replied:
Fixed.

artwyman commented on packages/lib/gpcircuits/test/proto-pod-gpc.spec.ts:
```diff
@@ -512,9 +612,13 @@ describe("proto-pod-gpc.ProtoPODGPC (Precompiled Artifacts) should work", functi
   ): CircuitArtifactPaths {
     const circuitDesc = ProtoPODGPC.pickCircuit(params);
-    expect(circuitDesc).to.not.be.undefined;
+    //    expect(circuitDesc).to.not.be.undefined;
```
Go ahead and delete this if you're relying on the error below for debuggability.
When I wrote this, this line was what I wanted to check, and the if statement below was just to make the TypeScript compiler happy.

artwyman commented on packages/lib/gpcircuits/test/tuple.spec.ts:
```diff
@@ -0,0 +1,190 @@
+  });
+
+  it("should compute the right tuple indices for different input tuples and parameters", () => {
+    [
```
I like this pattern for iterating through test-cases.
One suggestion particularly for cases as large as this one is that it might be clearest to use objects with named fields rather than bare arrays.  That way you could have the 3 named args, and an `expectedResult`.

artwyman reviewed (changes requested):
Looks good!  Lots of comments, but they're mostly about clarity rather than the core algorithm, on which we're pretty well aligned.

I only glanced at the test code.  It looks like it probably needs more negative test cases.  Feel free to leave comments like `TODO(POD-P3)` for things you think are worth doing but might not get to in this PR.

ax0 commented on packages/lib/gpcircuits/src/list-membership.ts:
```diff
@@ -0,0 +1,183 @@
+  firstTupleIndex: number,
+  memberIndex: number[],
+  list: PODValue[][]
+): { tupleIndices: number[][]; memberIndex: number; membershipList: bigint[] } {
```
I was thinking when the time comes this could be chained with whatever else requires tuples provided the `firstTupleIndex` parameter is appropriately set  on each call (and all `tupleIndices` are concatenated in the right order). `processLists` will have to be modified for this.

artwyman replied:
That's clever.  Maybe calling it firstAvailableTupleIndex would help clarify that purpose?

Something you'll have to figure out as you add this in to the @pcd/gpc layer is how much the definition of these tuples is explicit (i.e. there's a tuple config object) or implicit (i.e. the list config contains an input specification which is a tuple).  There's a lot of complexity here, but it might mean that the tuple definitions can be entirely implicit.

ax0 commented on packages/lib/gpcircuits/circuits/proto-pod-gpc.circom:
```diff
@@ -16,137 +18,215 @@ include "owner.circom";
+  // Membership list.
+  signal input membershipList[MAX_LISTS][MAX_LIST_ENTRIES];
+
+  signal memberHash[MAX_LISTS];
```
I agree, but alas, the powers that be have this to say:
> error[TAC01]: An anonymous component cannot be used with operator ===

artwyman replied:
Interesting.  That rings a bell, so I think I ran into the same thing in some of the other circuits.
Well, the rest of the simplification should work, right?  One extra line for the final comparison, but you shouldn't need the extra arrays in order to pass inputs from one template to another.

ax0 commented on packages/lib/gpcircuits/scripts/gen-circuit-parameters.ts:
```diff
@@ -17,17 +17,19 @@ import { MAX_PARALLEL_PROMISES, clearDir } from "./util";
+  [1, 5, 6, 1, 10, 2, 2],
+  [3, 10, 8, 1, 10, 2, 2],
+  [3, 10, 8, 2, 20, 1, 4],
+  [5, 15, 9, 3, 20, 4, 3]
```
`yarn gen-test-artifacts` takes 3 minutes and `yarn test` takes 75 seconds on my machine. I'll drop the last case.

ax0 commented on packages/lib/gpcircuits/src/list-membership.ts:
```diff
@@ -0,0 +1,183 @@
+ * `list` a list of singletons.
+ *
+ * @param params parameters of the ProtoPODGPC the list is processed for
+ * @param firstTupleIndex the index of the first output tuple in the combined array of
```
I think we clarified this below, but for reference: It will be `params.maxEntries` the first time it is called. The value will be greater on subsequent calls assuming tuples have been 'allocated'.

ax0 commented on packages/lib/gpcircuits/src/tuple.ts:
```diff
@@ -0,0 +1,171 @@
+import { poseidon2, poseidon3, poseidon4 } from "poseidon-lite";
+import { padArray, toChunks } from "./util";
+
+const poseidon = [poseidon2, poseidon3, poseidon4];
```
Actually, I'm wondering if this limitation is really necessary. Should I just go ahead and include all the Poseidon instances? In any case, I'm adding a comment here. One of the tuple helpers already throws an error if the arity is outside of the range (and the other relevant one will too after this review).

artwyman replied:
The Zupass team is putting effort into reducing bundle size, and the constants for each poseidon size are a contributor to that (much less with poseidon-lite than it was with circomlibjs admittedly).  Rob just encoded some similar limitations in EdDSAPCD so I think we should stick to the same here until we need something larger.  There's certainly no need for a hash larger than the largest circuit we choose to generate.

ax0 commented:
I have incorporated pretty much all of your suggestions. A couple of points:

- The most drastic change is the introduction of the Circom template `MultiTupleModule` and the corresponding TypeScript code. I've moved all of the tuple helpers to `multituple.ts` and introduced a couple of new ones in `tuple.ts` and `multituple.ts` for testing purposes.
- The tuple and list membership modules have been effectively disabled in the `gpc` package by feeding in appropriate circuit parameters and default values for the corresponding signal inputs. This will be sorted out in the next PR.

artwyman commented on packages/lib/gpcircuits/test/multituple.spec.ts:
```diff
@@ -0,0 +1,258 @@
+    }
+  });
+});
+// TODO(POD-P3): Flesh out tests.
```
What more needs testing at this level?

artwyman commented on packages/lib/gpcircuits/test/multituple.spec.ts:
```diff
@@ -0,0 +1,258 @@
+    );
+    const maxTuples = 5;
+    for (const tuple of [
+      [0, 1],
```
Try computing some of these out of order, or with gaps rather than sequential indices.

artwyman commented on packages/lib/gpcircuits/test/multituple.spec.ts:
```diff
@@ -0,0 +1,258 @@
+  });
+
+  it("should produce expected output", async () => {
+    // The case of one tuple
```
Make this its own test case.  It makes it easier to read test failures.

artwyman commented on packages/lib/gpcircuits/test/multituple.spec.ts:
```diff
@@ -0,0 +1,258 @@
+import { PODValue, podValueHash } from "@pcd/pod";
+import { BABY_JUB_NEGATIVE_ONE } from "@pcd/util";
+
+describe("MultiTuple helpers should work", function () {
```
Kudos for thorough tests of the helpers. I came to check as I read the code for them.  It's easier to evaluate coverage of edge cases in tests than by code inspection.

artwyman commented on packages/lib/gpcircuits/test/multituple.spec.ts:
```diff
@@ -0,0 +1,258 @@
+describe("MultiTuple helpers should work", function () {
+  it("should compute the right number of required tuples for different input tuple arities", () => {
+    [
+      [2, 2, 1],
```
There should be some cases returning 0?

artwyman commented on packages/lib/gpcircuits/test/multituple.spec.ts:
```diff
@@ -0,0 +1,258 @@
+
+  it("should compute the right maximum tuple arity representable by different parameters", () => {
+    [
+      [1, 2, 2],
```
There should be some cases returning 0?

artwyman commented on packages/lib/gpcircuits/src/multituple.ts:
```diff
@@ -0,0 +1,232 @@
+    );
+  }
+
+  // Note the first index.
```
Why?  I assume for padding later.
If it doesn't say why, then this comment doesn't add anything which the code doesn't already say.

artwyman commented on packages/lib/gpcircuits/src/multituple.ts:
```diff
@@ -0,0 +1,232 @@
+  // Note the first index.
+  const firstIndex = indices[0];
+
+  // The tuple indices will begin with the first `paramTupleArity` indices.
```
because the first tuple doesn't need to chain from a prior tuple

artwyman commented on packages/lib/gpcircuits/src/multituple.ts:
```diff
@@ -0,0 +1,232 @@
+    );
+  }
+
+  // Take note of the last tuple for padding purposes.
```
These variables, plus firstIndex above, could all fall below the single comment on the loop below.  That comment explains the purpose, and the variable names are very clear on each value's purpose without the need for extra comments about "noting" the values.

artwyman commented on packages/lib/gpcircuits/src/multituple.ts:
```diff
@@ -0,0 +1,232 @@
+  paramTupleArity: number,
+  firstAvailableTupleIndex: number,
+  indices: number[]
+): number[][] {
```
Also a lot clearer than before, thanks.

artwyman commented on packages/lib/gpcircuits/src/multituple.ts:
```diff
@@ -0,0 +1,232 @@
+}
+
+/**
+ * This procedure takes an N-tuple `elements` and computes its 'multituple hash',
```
Reorder this a bit to separate "what" from "how".  I think the last sentence of this paragraph is the one which belongs first, stating what this function is for.  The rest of the details could be their own paragraph afterwards, or they could be inside the function.  (For this kind of low-level calculation function, I think explaining the algorithm in the doc comment makes sense, but it should come after a description of how to use it.)

artwyman commented on packages/lib/gpcircuits/src/multituple.ts:
```diff
@@ -0,0 +1,232 @@
+ * @returns the list of multituple hashes
+ * @throws RangeError if `paramTupleArity` is outside the admissible range
+ */
+export function multiTupleHasher(
```
The structure of this function is sensible, though do you anticipate it being used anywhere outside of hashTuple?  I could see it being helpful to be able to calculate the expected output values of all the tuples in a module, but right now the code doesn't make use of that, instead making separate calls to hashTuple where necessary.

ax0 replied:
I suppose not. Its main purpose is to check the output of the multituple module, since the only use we currently have for chains of tuples is reducing a longer tuple down to a single hash for list membership checks, in which case `hashTuple` suffices as you noted.

artwyman replied:
That also begs the question of whether there are additional test cases to write based on that. :)

Fine to leave it as-is if it's potentially useful in future, particularly since it could come up as you implement the GPC config layer.

artwyman commented on packages/lib/gpcircuits/src/multituple.ts:
```diff
@@ -0,0 +1,232 @@
+  }
+
+  // Else, proceed as follows:
+  const hasher = (tuple: bigint[], hashes: bigint[]): bigint[] =>
```
I think I follow this (and further clarification is optional), but I think it could be clearer as a for loop rather than recursion.  I'm imagining `for (i = paramTupleArity; i += paramTupleArity; i < valueHashes.length)` to step appropriately, with the final partial tuple being built after the loop.

artwyman commented on packages/lib/gpcircuits/src/multituple.ts:
```diff
@@ -0,0 +1,232 @@
+        // chunk hash, pad and hash, appending it to the list of `hashes` computed
+        // so far.
+        hashes.concat(
+          tupleHasher(
```
This split of multi-tuple from tuple, mirroring the structure of the cirom templates, helps make this whole process easier to understand.  While analyzing this function I don't need to worry (yet) about the details of how `tupleHasher` does its job.

artwyman commented on packages/lib/gpcircuits/src/proto-pod-gpc.ts:
```diff
@@ -41,6 +41,13 @@ export type ProtoPODGPCInputs = {
   /*PUB*/ ownerExternalNullifier: CircuitSignal;
   /*PUB*/ ownerIsNullfierHashRevealed: CircuitSignal;
 
+  // Tuple module (1)
```
MultiTuple module.

artwyman commented on packages/lib/gpcircuits/src/proto-pod-gpc.ts:
```diff
@@ -41,6 +41,13 @@ export type ProtoPODGPCInputs = {
+  // Tuple module (1)
+  /*PUB*/ tupleIndices: CircuitSignal /*MAX_TUPLES*/[] /*TUPLE_ARITY*/[];
+
+  // List membership module (1)
```
1+ (since these are inputs to multiple modules)

artwyman commented on packages/lib/gpcircuits/src/proto-pod-gpc.ts:
```diff
@@ -184,7 +245,15 @@ export function protoPODGPCCircuitParamArray(
 ): ProtoPODGPCCircuitParams {
-  return ProtoPODGPCCircuitParams(params[0], params[1], params[2]);
+  return ProtoPODGPCCircuitParams(
+    params[0],
```
Optional, but I think `...params` might work here.  A similar thing would work in Python.

ax0 replied:
TypeScript complains if I do this unless I change `ProtoPODGPCCircuitParams` function signature to take a rest parameter, which will probably make things less pretty.

artwyman commented on packages/lib/gpcircuits/test/proto-pod-gpc.spec.ts:
```diff
@@ -400,12 +439,69 @@ function makeTestSignals(
+  ]
+    .map((pair) => pair as [number, PODValue[]])
+    // Consider only those entry values that are enabled.
+    .filter((pair) => sigEntryIsValueEnabled[pair[0]] === 1n)
```
I think this is unnecessary.  sigEntryIsValueEnabled indicates that the plaintext value is in the circuit, which only works for numbers.  Tuples and memberrship checks should work for strings and other values too.

artwyman commented on packages/lib/gpcircuits/test/proto-pod-gpc.spec.ts:
```diff
@@ -400,12 +439,69 @@ function makeTestSignals(
+      ? [
+          [1],
+          [sigEntryValue[1], 19n, 7n, 0n, 12n, 13n].map((value) => {
+            return [{ type: "cryptographic", value }];
```
Maybe use list2 to test a non-numeric value.

artwyman commented on packages/lib/gpcircuits/test/tuple.spec.ts:
```diff
@@ -0,0 +1,54 @@
+    await circuit.expectPass(sampleInput, sampleOutput);
+  });
+});
+// TODO(POD-P3): Flesh out tests.
```
What more is needed?  I think it's probably simple enough to do now for this module.

ax0 replied:
Added tests for tuple arities 2-4.

artwyman reviewed (changes requested):
Much clearer, thank you!  Almost all of my remaining comments are further clarity/doc improvements, and a few notes about tests (which I still skimmed, but more than last time).
The one functional note I recall is that the tuple module seems to allow for -1 indexes, but the TS code never makes use of that.  Probably that should be removed from the circom code since it's not being used for padding.

ax0 commented on packages/lib/gpcircuits/src/list-membership.ts:
```diff
@@ -0,0 +1,200 @@
+ * applications of {@link processSingleList}.
+ *
+ * @param params parameters of the ProtoPODGPC the list is processed for
+ * @param listComparisonValueIndices the indices of the entry values (or tuples) each of which is a member of the corresponding list
```
I suppose I wasn't clear that it represents either a single index or a list of indices representing the tuple that has to be formed to carry out the list membership check.

artwyman commented on packages/lib/gpcircuits/src/list-membership.ts:
```diff
@@ -103,13 +112,19 @@ export function processLists(
 /**
+ * Generates part of the ProtoPODGPC circuit inputs for a single
+ * list membership check. This is used to generate the full inputs
+ * for multiple list membersh> checks in (@link processLists).
```
```suggestion
 * for multiple list membership checks in (@link processLists).
```

artwyman commented on packages/lib/gpcircuits/test/tuple.spec.ts:
```diff
@@ -1,54 +1,68 @@
-    TupleHasherInputNamesType,
-    TupleHasherOutputNamesType
-  >;
+  const circuit: (
```
Particularly given the need to duplicate the types, I think this would probably be clearer as a `function` declaration rather than a `const`

ax0 replied:
Ah, the type duplication was unnecessary as it turns out. I had it in there back when the definition was split and forgot to get rid of it.

artwyman reviewed (approved):
Nice!  Ship it.

![launch-of-the-uss-detroit-nonononoyes](https://github.com/proofcarryingdata/zupass/assets/1761206/4999ea99-03ae-4005-b61d-f4441a78cc5d)
